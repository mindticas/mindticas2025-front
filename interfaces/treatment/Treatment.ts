export interface Treatment  {
    id: number;
    name: string;
    price: number;
    duration: number;
    description: string;
}