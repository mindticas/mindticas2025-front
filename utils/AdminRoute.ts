import Layout from '@/components/Layout';
import { Box } from '@chakra-ui/react';

export class AdminRoute {
    // Fixed route for admin
    private static route: string = '/admin';

    // Returns the Layout component if the pathname contains '/admin', otherwise returns box
    static getLayout(pathname: string) {
        return pathname.includes(this.route) ? Layout : Box;
    }
}
