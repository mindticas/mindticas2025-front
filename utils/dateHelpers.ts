// Gets the name of the day according to the full calendar format
export const getDayName = (dateStr: string): string => {
    const [year, month, day] = dateStr.split('-').map(Number)
    const date = new Date(year, month -1, day)
    return date.toLocaleDateString('es-ES', { weekday: 'long'})
}
// Get the milliseconds in a day
export function addDays(date: Date, days: number): Date {
    const milisecondsPerDay = 24 * 60 * 60 * 1000; 
    return new Date(date.getTime() + days * milisecondsPerDay)
}