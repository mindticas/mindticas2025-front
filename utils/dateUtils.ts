import { DateTime } from 'luxon';

interface DateTimeParts {
    date: string;
    time: string;
}

// * @returns Object with (YYYY-MM-DD) y time (HH:MM)

export const splitDateTimeFromISO = (
    isoString: string | null | undefined,
): DateTimeParts => {
    if (!isoString) return { date: '', time: '' };

    const dateObj = new Date(isoString);

    if (isNaN(dateObj.getTime())) {
        return { date: '', time: '' };
    }

    // format date (YYYY-MM-DD)
    const date = DateTime.fromJSDate(dateObj)
        .setZone('local')
        .toFormat('yyyy-MM-dd');

    // format time (HH:MM)
    const time = dateObj
        .toLocaleTimeString('es-ES', {
            hour: '2-digit',
            minute: '2-digit',
            hour12: false,
        })
        .replace(/\./g, '');

    return { date, time };
};
// Extract only the date in ISO format.
export const extractDateOnly = (isoDate: string | null | undefined): string => {
    if (!isoDate) return '';
    return isoDate.split('T')[0];
};
// Check if a specific date is within the date range
export const isDateInRange = (
    dateToCheck: string | null | undefined,
    startDate: string | null | undefined,
    endDate: string | null | undefined,
): boolean => {
    if (!dateToCheck || !startDate || !endDate) return false;
    const dateOnly = extractDateOnly(dateToCheck);
    const startOnly = extractDateOnly(startDate);
    const endOnly = extractDateOnly(endDate);

    return dateOnly >= startOnly && dateOnly <= endOnly;
};
