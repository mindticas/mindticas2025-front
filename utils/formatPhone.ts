export default function formatPhone(phone: string): string {
    // Remove everything that is not digits
    const digits = phone.replace(/\D/g, '');
    // Applies format only if it has 10 digits
    if (digits.length === 10) {
        return `${digits.slice(0, 3)}-${digits.slice(3, 6)}-${digits.slice(6)}`;
    }
    return phone; // Returns the original if it doesn't match
}
