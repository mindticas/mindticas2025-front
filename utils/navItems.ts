import { BarChart, Calendar, Scissors, UserCog, Users } from 'lucide-react';

export const navItems = [
    { name: 'Citas', href: '/admin/', icon: Calendar },
    { name: 'Clientes', href: '/admin/clients', icon: Users },
    { name: 'Empleados', href: '/admin/employees', icon: UserCog },
    { name: 'Tratamientos', href: '/admin/treatments', icon: Scissors },
    { name: 'Reportes', href: '/admin/reports', icon: BarChart },
];
