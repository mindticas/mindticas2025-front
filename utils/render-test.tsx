import { Provider } from '@/components/ui/provider';
import { BookingProvider } from '@/context/BookingContext';
import { render as rtlRender } from '@testing-library/react';

export function render(ui: React.ReactNode) {
    return rtlRender(<>{ui}</>, {
        wrapper: (props: React.PropsWithChildren) => (
            <Provider>
                <BookingProvider>{props.children}</BookingProvider>
            </Provider>
        ),
    });
}
