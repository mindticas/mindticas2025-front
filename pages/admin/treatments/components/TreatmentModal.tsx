import ErrorMessage from '@/components/ErrorMessage';
import { Treatment } from '@/interfaces/treatment/Treatment';
import { validateTreatmentFields } from '@/utils/treatments/treatmentValidation';
import {
    Box,
    Button,
    CloseButton,
    Dialog,
    DialogPositioner,
    Field,
    Input,
    Portal,
    Text,
} from '@chakra-ui/react';
import { useEffect, useState } from 'react';

interface TreatmentModalProps {
    isOpen: boolean;
    onClose: () => void;
    mode: 'create' | 'edit' | 'cancel' | 'delete';
    selectedTreatment?: Treatment;
    onTreatmentCreated?: () => void;
    onSubmit: (treatmentData?: Omit<Treatment, 'id'>) => Promise<void>;
    isSubmitting: boolean;
}

export default function TreatmentModal({
    isOpen,
    onClose,
    mode,
    selectedTreatment,
    onSubmit,
    isSubmitting,
}: TreatmentModalProps) {
    const [formData, setFormData] = useState({
        name: '',
        description: '',
        price: 0,
        duration: 0,
    });
    const [errors, setErrors] = useState<Record<string, string>>({});

    useEffect(() => {
        if (isOpen) {
            setErrors({});
        }
        if (isOpen && mode === 'edit' && selectedTreatment) {
            setFormData({
                name: selectedTreatment.name,
                description: selectedTreatment.description,
                price: selectedTreatment.price,
                duration: selectedTreatment.duration,
            });
            return;
        }
        if (mode === 'create') {
            setFormData({
                name: '',
                description: '',
                price: 0,
                duration: 0,
            });
        }
    }, [isOpen, mode, selectedTreatment]);

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value, type } = e.target;
        // clean errors
        setErrors((prevErrors) => {
            const newErrors = { ...prevErrors };
            delete newErrors[name];
            return newErrors;
        });
        setFormData({
            ...formData,
            [name]: type === 'number' ? Number(value) : value,
        });
    };

    const handleSubmit = () => {
        if (mode === 'create' || mode === 'edit') {
            // Field validation
            const fieldErrors = validateTreatmentFields(formData);
            if (Object.keys(fieldErrors).length > 0) {
                setErrors(fieldErrors);
                return;
            }
            onSubmit(formData);
        } else if (mode === 'delete') {
            onSubmit();
        }
    };

    const getTitle = () => {
        switch (mode) {
            case 'create':
                return 'Nuevo tratamiento';
            case 'edit':
                return 'Editar tratamiento';
            case 'delete':
                return 'Eliminar tratamiento';
            default:
                return 'Tratamiento';
        }
    };

    return (
        <Dialog.Root
            open={isOpen}
            onOpenChange={({ open }) => !open && onClose()}
            placement='center'
        >
            <Portal>
                <Dialog.Backdrop />
                <DialogPositioner>
                    <Dialog.Content p={5} backgroundColor='white'>
                        <Dialog.Header
                            mb={4}
                            display='flex'
                            justifyContent='space-between'
                            alignContent='center'
                        >
                            <Dialog.Title>{getTitle()}</Dialog.Title>
                            <Dialog.CloseTrigger asChild>
                                <CloseButton color='black' size='sm' />
                            </Dialog.CloseTrigger>
                        </Dialog.Header>

                        <Dialog.Body>
                            {mode === 'create' || mode === 'edit' ? (
                                <form>
                                    <Box mb={4}>
                                        <Field.Root>
                                            <Field.Label fontWeight='semibold'>
                                                Nombre
                                            </Field.Label>
                                            <Input
                                                p={2}
                                                name='name'
                                                placeholder='Nombre del tratamiento'
                                                value={formData.name}
                                                onChange={handleInputChange}
                                                required
                                            />
                                            <ErrorMessage
                                                message={errors.name}
                                            />
                                        </Field.Root>
                                    </Box>
                                    <Box mb={4}>
                                        <Field.Root>
                                            <Field.Label fontWeight='semibold'>
                                                Descripción
                                            </Field.Label>
                                            <Input
                                                p={2}
                                                name='description'
                                                placeholder='Descripción'
                                                value={formData.description}
                                                onChange={handleInputChange}
                                                required
                                            />
                                            <ErrorMessage
                                                message={errors.description}
                                            />
                                        </Field.Root>
                                    </Box>
                                    <Box mb={4}>
                                        <Field.Root>
                                            <Field.Label fontWeight='semibold'>
                                                Precio
                                            </Field.Label>
                                            <Input
                                                type='number'
                                                p={2}
                                                name='price'
                                                placeholder='Precio'
                                                value={formData.price}
                                                onChange={handleInputChange}
                                                required
                                            />
                                            <ErrorMessage
                                                message={errors.price}
                                            />
                                        </Field.Root>
                                    </Box>
                                    <Box mb={4}>
                                        <Field.Root>
                                            <Field.Label fontWeight='semibold'>
                                                Duración
                                            </Field.Label>
                                            <Input
                                                type='number'
                                                p={2}
                                                name='duration'
                                                placeholder='Duración'
                                                value={formData.duration}
                                                onChange={handleInputChange}
                                                required
                                            />
                                            <ErrorMessage
                                                message={errors.duration}
                                            />
                                        </Field.Root>
                                    </Box>
                                </form>
                            ) : mode === 'delete' ? (
                                <Text>
                                    ¿Estás seguro que deseas eliminar el
                                    tratamiento "{selectedTreatment?.name}" Esta
                                    acción no se puede deshacer.
                                </Text>
                            ) : (
                                <></>
                            )}
                        </Dialog.Body>
                        <Dialog.Footer p={4}>
                            <Button
                                backgroundColor={
                                    mode === 'delete' ? 'black' : 'red'
                                }
                                p={3}
                                color='white'
                                onClick={() => {
                                    onClose();
                                    setErrors({});
                                }}
                            >
                                {mode === 'delete' ? 'Cancelar' : 'Cancelar'}
                            </Button>
                            <Button
                                backgroundColor={
                                    mode === 'delete' ? 'red' : '#1C4ED8'
                                }
                                p={3}
                                color='white'
                                onClick={handleSubmit}
                                loading={isSubmitting}
                            >
                                {mode === 'create'
                                    ? 'Crear'
                                    : mode === 'edit'
                                    ? 'Actualizar'
                                    : 'Eliminar'}
                            </Button>
                        </Dialog.Footer>
                    </Dialog.Content>
                </DialogPositioner>
            </Portal>
        </Dialog.Root>
    );
}
