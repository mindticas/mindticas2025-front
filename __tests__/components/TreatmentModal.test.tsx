import TreatmentModal from '@/pages/admin/treatments/components/TreatmentModal';
import { Treatment } from '@/interfaces/treatment/Treatment';
import { render } from '@/utils/render-test';
// Treatment mockup for testing
const mockTreatment: Treatment = {
    id: 1,
    name: 'Corte de pelo',
    description: 'Corte básico',
    price: 150,
    duration: 30,
};

describe('TreatmentModal Snapshot Tests', () => {
    const mockOnClose = jest.fn();
    const mockOnSubmit = jest.fn();

    it('should match create mode snapshot', () => {
        const { asFragment } = render(
            <TreatmentModal
                isOpen={true}
                onClose={mockOnClose}
                mode='create'
                onSubmit={mockOnSubmit}
                isSubmitting={false}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should match edit mode snapshot', () => {
        const { asFragment } = render(
            <TreatmentModal
                isOpen={true}
                onClose={mockOnClose}
                mode='edit'
                selectedTreatment={mockTreatment}
                onSubmit={mockOnSubmit}
                isSubmitting={false}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should match delete mode snapshot', () => {
        const { asFragment } = render(
            <TreatmentModal
                isOpen={true}
                onClose={mockOnClose}
                mode='delete'
                selectedTreatment={mockTreatment}
                onSubmit={mockOnSubmit}
                isSubmitting={false}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with errors', () => {
        // Mock validateTreatmentFields to force errors
        jest.mock('@/utils/treatments/treatmentValidation', () => ({
            validateTreatmentFields: () => ({
                name: 'Nombre requerido',
                price: 'Precio inválido',
            }),
        }));

        const { asFragment } = render(
            <TreatmentModal
                isOpen={true}
                onClose={mockOnClose}
                mode='create'
                onSubmit={mockOnSubmit}
                isSubmitting={false}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot when submitting', () => {
        const { asFragment } = render(
            <TreatmentModal
                isOpen={true}
                onClose={mockOnClose}
                mode='edit'
                selectedTreatment={mockTreatment}
                onSubmit={mockOnSubmit}
                isSubmitting={true}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should match closed modal snapshot', () => {
        const { asFragment } = render(
            <TreatmentModal
                isOpen={false}
                onClose={mockOnClose}
                mode='create'
                onSubmit={mockOnSubmit}
                isSubmitting={false}
            />,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
