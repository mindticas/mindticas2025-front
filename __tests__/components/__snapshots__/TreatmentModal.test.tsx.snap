// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreatmentModal Snapshot Tests should match closed modal snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;

exports[`TreatmentModal Snapshot Tests should match create mode snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;

exports[`TreatmentModal Snapshot Tests should match delete mode snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;

exports[`TreatmentModal Snapshot Tests should match edit mode snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;

exports[`TreatmentModal Snapshot Tests should match snapshot when submitting 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;

exports[`TreatmentModal Snapshot Tests should match snapshot with errors 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</DocumentFragment>
`;
